### DISCUSSION TITLE
`SpatieMediaLibraryImageColumn` is empty when using `$this->form->fill(...)`

### DISCUSSION THREAD

**Discussion by bilogic** [filamentphp/filament#8166] [Help] [answered]
1. I have a `StaffResource` that has a `Image` field typed as `SpatieMediaLibraryImageColumn`
2. When editing the `StaffResource`, `Image` loads correctly
3. I also needed a custom `Page` where the staff can edit their own profile, which I used the following
    ```PHP
    public function mount(): void
    {
        $staff = Staff::find($id);

        dd($staff->toArray()); // `image` has the value of [] here...

        $this->form->fill($staff->toArray());
    }
    ```
4. I tried looking into `EditStaff.php` but was not able to gain any insights.
5. How do I make the profile page load image like the `EditStaff` page? Problem seems to be that `$staff->toArray()` returns an empty `Image`

**Comment by danharrin** [comment]
Have you passed the staff to the ->model() of the form https://filamentphp.com/docs/3.x/forms/adding-a-form-to-a-livewire-component#setting-a-form-model

**Reply by bilogic** [reply]
Thanks for getting back. I just did, but photo still doesn't show, see https://gist.github.com/bilogic/439e7634124a6aa9626c3e24236a9796#file-profile-php

**Reply by danharrin** [reply]
Are you sure there is actually an image in that collection belonging to that staff member

**Reply by bilogic** [reply]
yes, because I have another `StaffResource` filament resource and it is able to show the image for every staff under the `staff-image` collection

In `StaffResource.php`
```PHP
    public static function form(Form $form): Form
    {
        return $form
            ->schema([
                SpatieMediaLibraryFileUpload::make('Image')
                    ->collection('staff-image')
                    ->image()
                    ->reorderable()
                    ->openable()
                    ->downloadable()
                    ->imageEditor(),
                TextInput::make('name'),
            ]);
    }
```

**Reply by bilogic** [reply]
> ```PHP
> class Test extends Page implements HasForms
> {
>    use InteractsWithForms;>    
>    ...
>    public function mount(): void
>    {
>        $staff = Staff::find($id);
>    
>        dd($staff->toArray()); // `image` has the value of [] here...
>    
>        $this->form->fill($staff->toArray());
>    }
> ```

Could this be the clue? Because when I `dd()` in `EditStaff.php`, the `image` key has content, not just `[]`

**Reply by danharrin** [reply]
You should not have a column in your database called `image`. Spatie's package doesn't use that.

**Reply by bilogic** [reply]
> You should not have a column in your database called `image`. Spatie's package doesn't use that.

I don't have a `image` field in my table if that is what you mean. I only have `getImageAttribute()` method in my `Staff` model. 

Again, I would like to highlight, using `make:filament-resource Staff`, everything works. 
However, on a custom page trying to host just the edit form, existing image does not load/show.

**Reply by danharrin** [reply]
Please remove the getImageAttribute(), it could be clashing.

**Reply by danharrin** [reply]
The `image` key should not be in the `toArray()`, that is a problem I think

**Reply by bilogic** [reply]
Hmm, ok I removed `getImageAttribute` and reloaded the custom page, `name` was there, but still no `image`.

I think we should look at the bigger picture.

1. `StaffResource.php` works correctly with `getImageAttribute`, so why would it clash?
2. If I were to XDebug `EditStaff.php::mount()`, `image` is there, so why is it a problem that `toArray()` has `image`?

I can't figure out what I'm missing, but based on comparing with filament generated code, I'm quite certain it isn't `getImageAttribute` nor `toArray()` having `image`.

**Reply by danharrin** [reply]
Before `return $form`, `dd($this->staff)`. Is it null there

**Reply by bilogic** [reply]
Yes, it is. `dd($this->staff)` shows `image` with `null`

![image](https://github.com/filamentphp/filament/assets/946010/816dbcee-79f5-4565-b096-ecd9a60924ea)

I suppose that was because `->toArray()` was already `null` at `$this->form->fill($this->staff->toArray());`

So perhaps the question is, what did `EditStaff.php` do to make `toArray()`'s `image` not null

**Reply by bilogic** [reply]
https://github.com/filamentphp/spatie-laravel-media-library-plugin/blob/5e67253a4aad0195fef44be803ede1096981056c/src/Forms/Components/SpatieMediaLibraryFileUpload.php#L48
Above is the bit that populates `image`, basically by calling `$model->getMedia();`

Further, I performed a comparison by doing a `dd($this->data)` in `render()` of `EditStaff.php` and `Test.php`

![image](https://github.com/filamentphp/filament/assets/946010/eb1e3a2f-38c7-4a40-bfdb-cf269f441a22)

![image](https://github.com/filamentphp/filament/assets/946010/b8a9b0e1-33ec-44fd-a810-60a0f4a2539a)

However, the `Test.php` continues to not load the `Image`... one works while the other doesn't... really pulling my hair on this

**Reply by danharrin** [reply]
Are you absolutely sure that you dont have a column called `image` in your DB table, I dont understand why its in your model as dynamic attributes shouldnt be there by default

**Reply by bilogic** [reply]
Ok, you seem to be saying that if `image` field exists in my table, `SpatieMediaLibraryFileUpload` will not work, why does it work in `EditStaff.php` then? Are we chasing the right concern here?

Nonetheless, I have double checked my migrations again and indeed found that `image` field exists. However, after removing it, `Test.php` still does not show the image in `SpatieMediaLibraryFileUpload`.

This is a `dd($this->data)` in `Test.php::render()`

![image](https://github.com/filamentphp/filament/assets/946010/90d88e31-3642-4059-bd38-752faf35f361)

**Reply by bilogic** [reply]
@danharrin I managed to reproduce the issue here https://github.com/bilogic/filament-spatie-upload-issue, appreciate your help and insights. Thank you.

### Staff's image can be seen when editing
![image](https://github.com/filamentphp/filament/assets/946010/48fe474b-9625-452f-b8cc-dd33c9566ec8)

### Test custom page shows only the Name and no image
![image](https://github.com/filamentphp/filament/assets/946010/6eae18f9-45ba-4a8d-84ba-d246c18a8ac8)

**Reply by danharrin** [reply]
Please open an issue

**Reply by danharrin** [reply, marked as answer]
Having a quick look, I think the problem is `protected $staff;` needs to be `public $staff;`, otherwise Livewire won't persist it.

**Reply by bilogic** [reply]
> Having a quick look, I think the problem is `protected $staff;` needs to be `public $staff;`, otherwise Livewire won't persist it.

No, that did not solve it. I mean, if you refer to my screenshot of `Test.php`, `name` was persisted, only `image` was not. If `protected` was the cause, `name` should also be empty. I will open an issue. Thank you.

**Reply by danharrin** [reply]
Actually, name would not be affected as it is stored in a public array. Images are not loaded on the first request to the server AFAIK which is why it needs to be public

**Reply by bilogic** [reply]
I tried making `$staff` public again and this time it worked. Not sure what I did wrongly yesterday. Thank you.
